action/protocol/poll/protocol_test.go

```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
+	bd := &action.EnvelopeBuilder{}
+	elp := bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(tsf).Build()
+	selp, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp)
+	// Case 1: wrong action type
+	err=p.Validate(ctx,selp.Action())
+	require.Nil(err)
+	// Case 2: Only producer could create this protocol
+	p2,ctx2,ws2,_:=initConstruct(t)
+	require.NoError(p2.Initialize(ctx2, ws2))
+	var sc2 state.CandidateList
+	require.NoError(ws2.State(candidatesutil.ConstructKey(1), &sc2))
+	act2 := action.NewPutPollResult(1, 1, sc2)
+	elp = bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(act2).Build()
+	selp2, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp2)
+	caller, err := address.FromBytes(selp.SrcPubkey().Hash())
+	require.NoError(err)
+	ctx2 = protocol.WithValidateActionsCtx(
+		context.Background(),
+		protocol.ValidateActionsCtx{
+			BlockHeight:  1,
+			ProducerAddr: recipientAddr.String(),
+			Caller:       caller,
+		},
+	)
+	err=p.Validate(ctx2,selp2.Action())
+	require.True(strings.Contains(err.Error(), "Only producer could create this protocol"))
+	// Case 3: duplicate candidate
+	p3,ctx3,ws3,_:=initConstruct(t)
+	require.NoError(p3.Initialize(ctx3, ws3))
+	var sc3 state.CandidateList
+	require.NoError(ws3.State(candidatesutil.ConstructKey(1), &sc3))
+	sc3=append(sc3,&state.Candidate{"1",big.NewInt(10),"2"})
+	sc3=append(sc3,&state.Candidate{"1",big.NewInt(10),"2"})
+	act3 := action.NewPutPollResult(1, 1, sc3)
+	elp = bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(act3).Build()
+	selp3, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp3)
+	ctx3 = protocol.WithValidateActionsCtx(
+		context.Background(),
+		protocol.ValidateActionsCtx{
+			BlockHeight:  1,
+			ProducerAddr: testaddress.Addrinfo["producer"].String(),
+			Caller:       caller,
+		},
+	)
+	err=p.Validate(ctx3,selp3.Action())
+	require.True(strings.Contains(err.Error(), "duplicate candidate"))
+
+	// Case 4: delegate's length is not equal
+	p4,ctx4,ws4,_:=initConstruct(t)
+	require.NoError(p4.Initialize(ctx4, ws4))
+	var sc4 state.CandidateList
+	require.NoError(ws4.State(candidatesutil.ConstructKey(1), &sc4))
+	sc4=append(sc4,&state.Candidate{"1",big.NewInt(10),"2"})
+	act4 := action.NewPutPollResult(1, 1, sc4)
+	bd4 := &action.EnvelopeBuilder{}
+	elp4 := bd4.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(act4).Build()
+	selp4, err := action.Sign(elp4, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp4)
+	ctx4 = protocol.WithValidateActionsCtx(
+		context.Background(),
+		protocol.ValidateActionsCtx{
+			BlockHeight:  1,
+			ProducerAddr: testaddress.Addrinfo["producer"].String(),
+			Caller:       caller,
+		},
+	)
+	err=p4.Validate(ctx4,selp4.Action())
+	require.True(strings.Contains(err.Error(), "the proposed delegate list length"))
+
```

Comment:
add a case that the candidate's weight or something else are not identical

Reply:
ok